src/entities/pair.ts

```diff
@@ -106,6 +107,61 @@ export class Pair {
     return token.equals(this.token0) ? this.reserve0 : this.reserve1
   }
 
+  /**
+   * getAmountOut is the linear algebra of reserve ratio against amountIn:amountOut.
+   * https://ethereum.stackexchange.com/questions/101629/what-is-math-for-uniswap-calculates-the-amountout-and-amountin-why-997-and-1000
+   * has the math deduction for the reserve calculation without fee-on-transfer fees.
+   *
+   * With fee-on-transfer fees, intuitively it's just:
+   * inputAmountWithFeeAndTax = 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn
+   *                          = (1 - amountIn.sellFeesBips / 10000) * amountInWithFee
+   * outputAmountWithTax = amountOut * (1 - amountOut.buyFeesBips / 10000)
+   *
+   * But we are illustrating the math deduction below to ensure that's the case.
+   *
+   * before swap A * B = K where A = reserveIn B = reserveOut
+   *
+   * after swap A' * B' = K where only k is a constant value
+   *
+   * getAmountOut
+   *
+   * A' = A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn # here 0.3% is deducted
+   * B' = B - amountOut * (1 - amountOut.buyFeesBips / 10000)
+   * amountOut = (B - B') / (1 - amountOut.buyFeesBips / 10000) # where A' * B' still is k
+   *           = (B - K/(A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn))
+   *             /
+   *             (1 - amountOut.buyFeesBips / 10000)
+   *           = (B - AB/(A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn))
+   *             /
+   *             (1 - amountOut.buyFeesBips / 10000)
+   *           = ((BA + B * 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn - AB)/(A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn))
+   *             /
+   *             (1 - amountOut.buyFeesBips / 10000)
+   *           = (B * 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn / (A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn)
+   *             /
+   *             (1 - amountOut.buyFeesBips / 10000)
+   * amountOut * (1 - amountOut.buyFeesBips / 10000) = (B * 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn
+   *                                                    /
+   *                                                    (A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn)
+   *
+   * outputAmountWithTax = (B * 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn
+   *                       /
+   *                       (A + 0.997 * (1 - amountIn.sellFeesBips / 10000) * amountIn)
+   *                     = (B * 997 * (1 - amountIn.sellFeesBips / 10000) * amountIn
+   *                       /
+   *                       (1000 * A + 997 * (1 - amountIn.sellFeesBips / 10000) * amountIn)
+   *                     = (B * (1 - amountIn.sellFeesBips / 10000) * inputAmountWithFee)
```

Comment:
is that true? `amountIn` * 997 = inputAmountWithFee /1000 right?

Reply:
we should probably define `inputAmountWithFee` somewhere too

Reply:
I opt out to remove `inputAmountWithFee ` for `getAmountIn`, since the formula is not as straightforward as `getAmountOut`. I'm hoping to use the example from the unit tests I wrote to make it self-explanatory.